src/search3_query.erl

```diff
@@ -0,0 +1,40 @@
+-module(search3_query).
+
+-export([
+    run_query/4
+]).
+
+-include("search3.hrl").
+
+run_query(Db, DDoc, IndexName, QueryArgs) ->
+    #{name := DbName} = Db,
+    {ok, Index} = search3_util:design_doc_to_index(DDoc, IndexName),
+    Index1 = Index#index{dbname = DbName},
+    run_query(Db, Index1, QueryArgs).
+
+run_query(Db, Index, QueryArgs) ->
+    % A scenario exists wjere a pod dies at indexing time and the
+    % CommitedSeq is behind the UpdateSeq. This can be detected when the
+    % Session id after indexing is different than the session id of a query.
+    Session = maybe_build_index(Db, Index),
+    try
+        Response = search3_rpc:search_index(Index, QueryArgs),
+        search3_response:handle_search_response(Response, Session)
+    catch error:session_mismatch ->
+        run_query(Db, Index, QueryArgs)
+    end.
+
+maybe_build_index(Db, Index) ->
+    {Action, Session, WaitSeq} = fabric2_fdb:transactional(Db, fun(TxDb) ->
+        DbSeq = fabric2_db:get_update_seq(TxDb),
+        {UpdateSession, SearchSeq} = search3_rpc:get_update_seq(Index),
+        case DbSeq == SearchSeq of
```

Comment:
this is not how we usually test if an index needs updating.

If stale=ok, then we don't bother at all, but it seems we would index in that case? If stale=false, we fetch changes since SearchSeq and apply them. That list might be empty, but we don't infer that from a comparison of the two sequences at two different points in time.

I appreciate there's a txn around this, so DbSeq and SearchSeq are from a consistent fdb snapshot, and therefore testing for equality is fine, but it's another mechanism from the 'normal' one and that seems risky.